be/src/exec/write_combined_txn_log.cpp

```diff
@@ -24,4 +27,47 @@ Status write_combined_txn_log(const CombinedTxnLogPB& logs) {
     return tablet_mgr->put_combined_txn_log(logs);
 }
 
+void mark_failure(const Status& status, std::atomic<bool>* has_error, Status* final_status) {
+    if (!has_error->load()) {
+        if (!has_error->exchange(true)) {
+            if (final_status->ok()) {
+                *final_status = status;
+            }
+        }
+    }
+}
+
+std::function<void()> create_txn_log_task(const CombinedTxnLogPB* logs, lake::TabletManager* tablet_mgr,
+                                          std::atomic<bool>* has_error, Status* final_status) {
+    return [logs, tablet_mgr, has_error, final_status]() {
+        try {
+            Status status = tablet_mgr->put_combined_txn_log(*logs);
+            if (!status.ok()) {
+                throw std::runtime_error("Txn log write failed");
+            }
+        } catch (const std::exception& e) {
+            mark_failure(Status::IOError(e.what()), has_error, final_status);
+        } catch (...) {
+            mark_failure(Status::Unknown("Unknown exception in write combined txn log task"), has_error, final_status);
+        }
+    };
+}
+
+Status write_combined_txn_log(const std::map<int64_t, CombinedTxnLogPB>& txn_log_map) {
+    size_t task_count = txn_log_map.size();
+    CountDownLatch latch(task_count);
+    std::atomic<bool> has_error(false);
+    Status final_status;
+
+    for (const auto& [partition_id, logs] : txn_log_map) {
+        auto task_logic =
+                create_txn_log_task(&logs, ExecEnv::GetInstance()->lake_tablet_manager(), &has_error, &final_status);
+        auto task = std::make_shared<AutoCleanRunnable>(std::move(task_logic), [&latch]() { latch.count_down(); });
+        RETURN_IF_ERROR(ExecEnv::GetInstance()->put_combined_txn_log_thread_pool()->submit(task));
```

Comment:
This early return is dangerous. E.g. if the first task submit success and then the second fails and return here. The early return will cause the destruction of `has_error` and `latch`, where when the first task get executed, it manipulates on invalid object and crash.